cpp/account_server.cc

```diff
@@ -0,0 +1,128 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <unistd.h>
+#include <iostream>
+#include <memory>
+#include <string>
+
+#include <grpcpp/ext/proto_server_reflection_plugin.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/health_check_service_interface.h>
+
+#include "examples/protos/account.grpc.pb.h"
+
+using account::Account;
+using account::GetUserInfoRequest;
+using account::GetUserInfoResponse;
+using account::MembershipType;
+using grpc::Server;
+using grpc::ServerBuilder;
+using grpc::ServerContext;
+using grpc::Status;
+
+class AccountServiceImpl final : public Account::Service {
+ public:
+  void SetHostName(const std::string& hostname) { hostname_ = hostname; }
+
+ private:
+  Status GetUserInfo(ServerContext* context, const GetUserInfoRequest* request,
+                     GetUserInfoResponse* response) override {
+    std::string token = request->token();
+    context->AddInitialMetadata("hostname", hostname_);
+    if (token == "2bd806c9") {
+      response->set_name("Alice");
+      response->set_membership(MembershipType::PREMIUM);
+    } else if (token == "81b637d8") {
+      response->set_name("Bob");
+      response->set_membership(MembershipType::NORMAL);
+    } else {
+      std::cout << "Default to Bob" << std::endl;
+      response->set_name("Bob");
+      response->set_membership(MembershipType::NORMAL);
+    }
+    return Status::OK;
+  }
+
+  std::string hostname_;
+};
+
+void RunServer(const std::string& port, const std::string& hostname_suffix) {
+  char base_hostname[256];
+  if (gethostname(base_hostname, 256) != 0) {
+    std::cout << "unable to get host name" << std::endl;
+    return;
+  }
+  std::string hostname(base_hostname);
+  hostname += hostname_suffix;
+  std::string server_address("0.0.0.0:");
+  server_address += port;
+  AccountServiceImpl service;
+  service.SetHostName(hostname);
```

Comment:
Is it better to use a constructor instead of set? `AccountServiceImpl service(hostname)`?

Reply:
done